Refactor and relocate connect_to_tor

aagbsn opened the pull request:
this unifies some code and makes sure it is exercised by a unit test

aagbsn committed:
Refactor and relocate connect_to_tor

DonnchaC commented on test/template.py:
```diff
@@ -7,16 +7,15 @@
-        self.tor = yield build_tor_connection(
-            TCP4ClientEndpoint(reactor, 'localhost', int(
-                os.environ.get('CHUTNEY_CONTROL_PORT'))))
+        self.tor = yield connect_to_tor(False, 30, 30,
```
I would specify the key with the keyword arguments to make it clear what each value represents.

aagbsn replied:
done

aagbsn committed:
rename or_endpoint to tor_socks_endpoint

aagbsn committed:
consolidate options

aagbsn committed:
add errback

aagbsn committed:
use build_tor_connection instead

build_local_tor_connection tries to connect to a local listening
socket first, if one exists. If someone wants to use a locally
running tor instance other than the local system tor
 (i.e. uses chutney) this won't work.

aagbsn committed:
relocate config

aagbsn committed:
delay by sending get_conf

aagbsn committed:
dont get_conf

DonnchaC commented on bwscanner/attacher.py:
```diff
@@ -162,3 +147,46 @@ def add_single_port_exit(port):
+        'CircuitBuildTimeout': circuit_build_timeout,
+        'CircuitIdleTimeout': circuit_idle_timeout,
+        'UseEntryGuards': 0,  # Disable UseEntryGuards to avoid PathBias warnings.
+        'UseMicroDescriptors': 0,
```
Setting UseMicroDescriptors causes Tor to fetch new descriptors and Txtorcon then updates it's TorState object. There is a race where this state update can cause the tests to fail.

aagbsn committed:
wait for a NEWCONSENSUS after set_conf

Thanks to meejah for the example code.

TODO: refactor to only do this if we need to (apparently when toggling
UseMicroDescriptors)

aagbsn commented:
It looks like we fixed a few more bugs than this PR indicates. In particular:
1. on systems with a local system tor and control socket, build_local_tor_connection would pick the system-tor (this is wrapped in a try/except).

2. set_conf was used to set UseMicroDescriptors to 0, which caused a race between bwscanner and tor fetching more descriptors. Now we wait until a NEWCONSENSUS event has been emitted before proceeding.

aagbsn commented:
Closing this PR in favor of opening a new PR with DonnchaC's refactoring
